README.md
<!-- badges -->
[![Hex.pm Version](http://img.shields.io/hexpm/v/sanity_webhook_plug)](https://hex.pm/packages/sanity_webhook_plug)
[![Hex docs](http://img.shields.io/badge/hex.pm-docs-blue.svg?style=flat)](https://hexdocs.pm/sanity_webhook_plug)
[![License](https://img.shields.io/hexpm/l/sanity_webhook_plug)](./LICENSE)

# Sanity Webhook Plug

You're reading the main branch's readme. Please visit
[hexdocs](https://hexdocs.pm/sanity_webhook_plug) for the latest published documentation.

<!-- MDOC !-->

SanityWebhookPlug is a Plug that verifies Sanity webhooks for your Elixir Plug
application. Designed to work with [Sanity GROQ-powered
webhooks](https://www.sanity.io/docs/webhooks)

## Installation

```elixir
def deps do
  [
    {:sanity_webhook_plug, "~> 0.1.4"}
  ]
end
```

## Usage

Use this plug in your endpoint:

```elixir
# If using Plug or Phoenix, place before `plug Plug.Parsers`
# For Phoenix apps, in lib/my_app_web/endpoint.ex:
plug SanityWebhookPlug,
  at: "/webhooks/sanity",
  handler: MyAppWeb.SanityWebhookHandler
```

You may alternatively configure the secret in config, which will be read during
runtime:

```elixir
# in config/runtime.exs
config :sanity_webhook_plug,
  webhook_secret: System.get_env("SANITY_WEBHOOK_SECRET")
```

Define a handler to handle webhooks:

```elixir
defmodule MyAppWeb.SanityWebhookHandler do
  @behaviour SanityWebhookPlug.Handler
  alias Plug.Conn

  # see below for an example using Phoenix

  @impl SanityWebhookPlug.Handler
  def handle_event(conn, params) do
    # Process and return the conn

    conn
    |> Conn.put_resp_header("content-type", "application/json")
    |> Conn.send_resp(200, Jason.encode!(%{success: "yay!"}))
  end

  @impl SanityWebhookPlug.Handler
  def handle_error(conn, error) do
    # Process and return the conn

    conn
    |> Conn.put_resp_header("content-type", "application/json")
    |> Conn.send_resp(500, Jason.encode!(%{error: "uh oh!"}))
  end
end
```

### Options:

- `:at` (required): The request path to match against. eg, `"/webhooks/sanity"`
- `:handler` (required): The controller-like module that responds to
    `handle_event/2` that is passed the conn and the params, and
    `handle_error/2` that is passed the conn and the error. The error may be an
    exception or a string.
- `:secret`: The Sanity webhook secret. eg: `123abc`. Supplying an MFA tuple will
    be called at runtime, otherwise it will be compiled. If not set, it will
    obtain via `Application.get_env(:sanity_webhook_plug, :webhook_secret)`.
    If supplying an MFA or function reference, it must return `{:ok, my_secret}`
    or a string.
- `:json_decoder`: JSON encoding library. When not supplied, it will use choose
    Phoenix's configured library, `Jason`, or `Poison`. Sanity requires
    JSON-encoded responses.

Options forwarded to `Plug.Conn.read_body/2`:

- `:length` - sets the number of bytes to read from the request at a time.
- `:read_length` - sets the amount of bytes to read at one time from the
    underlying socket to fill the chunk.
- `:read_timeout` - sets the timeout for each socket read.

Verifying the signature requires reading the body, but its best to do this
before _interpreting_ the body into JSON or other parsed formats. Plug can
protect your system by limiting how much of body to read to prevent exhaustion.
Ideally, any of these settings you have for `Plug.Parsers` in your endpoint, you
should also have for SanityWebhookPlug.

The body and query params will be merged and given to your handler, which
matches Phoenix behavior.

### Example

An example using Phoenix

```elixir
## In lib/my_app_web/endpoint.ex

# place before Plug.Parsers
defmodule MyAppWeb.Endpoint do
  use Phoenix.Endpoint, otp_app: :my_app

  # ...

  plug SanityWebhookPlug,
    at: "/webhooks/sanity",
    handler: MyAppWeb.SanityWebhookHandler

  # Plug.Parsers down here somewhere
end


## in lib/my_app_web/controllers/sanity_webhook_handler.ex
def MyAppWeb.SanityWebhookHandler do
  use MyAppWeb, :controller
  require Logger
  @behaviour SanityWebhookPlug.Handler

  # handle known events
  def handle_event(conn, %{"_type" => type, "_id" => id}) do
    # do something
    json(conn, %{success: "Did the thing!"})
  end

  def handle_event(conn, params) do
    Logger.warn("SanityWebhook: unhandled webhook: #{inspect(params)}")

    conn
    |> put_status(500)
    |> json(%{error: "unhandled webhook"})
  end

  def handle_error(conn, error) do
    debug = SanityWebhookPlug.get_debug(conn)
    Logger.error("SanityWebhook error: #{inspect(debug)}")

    conn
    |> put_status(400)
    |> json(%{error: inspect(error)})
  end
end
```


CHANGELOG.md
# Changelog

## 0.1.4 (2023-03-24)

- Update project URLs

## 0.1.3 (2023-01-25)

- Correct documentation

## 0.1.2 (2023-01-23)

- Skip halted conns
- Add `make_header/3` function

## 0.1.1 (2023-01-09)

- Halt conn after successful event

## 0.1.0 (2023-01-06)

- Initial release :tada:


test/sanity_webhook_plug_test.exs
defmodule SanityWebhookPlugTest do
  use ExUnit.Case
  use Plug.Test
  import ExUnit.CaptureIO
  alias SanityWebhookPlug.Signature

  def secret, do: {:ok, "test"}
  def bare_secret, do: "test"
  def bad_secret, do: {:ok, "test2"}
  def error_secret, do: {:error, "no secret"}

  @test_payload File.read!("test/support/test.jpg")
  @good_payload Jason.encode!(%{"_id" => "resume"})
  @good_ts 1_633_519_811_129
  @good_hash "tLa470fx7qkLLEcMOcEUFuBbRSkGujyskxrNXcoh0N0"
  @good_signature "t=#{@good_ts},v1=#{@good_hash}"
  @opts [
    handler: __MODULE__.Handler,
    at: "/sanity",
    secret: "test"
  ]

  defmodule Handler do
    def handle_event(conn, _params) do
      Plug.Conn.send_resp(conn, 200, Jason.encode!(%{success: "ok"}))
    end

    def handle_error(conn, error) when is_binary(error) do
      Plug.Conn.send_resp(conn, 400, Jason.encode!(%{error: error}))
    end

    def handle_error(conn, error) when is_exception(error) do
      Plug.Conn.send_resp(conn, 400, Jason.encode!(%{error: "error"}))
    end
  end

  test "secret is not printed" do
    # credo:disable-for-lines:1
    refute capture_io(fn -> IO.inspect(%SanityWebhookPlug{secret: "foo"}) end) =~ "foo"
  end

  test "base64url encoding and decoding" do
    original = "ladies and gentlemen, we are floating in space"
    encoded = "bGFkaWVzIGFuZCBnZW50bGVtZW4sIHdlIGFyZSBmbG9hdGluZyBpbiBzcGFjZQ"
    assert Signature.base64url_encode(original) == encoded
    assert Signature.base64url_decode(encoded) == original

    bin = Signature.base64url_encode(@test_payload)
    refute String.contains?(bin, "+")
    refute String.contains?(bin, "/")
    refute String.contains?(bin, "=")
    assert Signature.base64url_decode(bin) == @test_payload
  end

  test "works with single path" do
    opts = Keyword.put(@opts, :path, "/sanity")

    conn =
      @good_payload
      |> setup_conn(@good_signature)
      |> call(opts)

    assert %{error: false} = conn.private.sanity_webhook_plug
  end

  test "skips without path match" do
    opts = Keyword.put(@opts, :path, "/no-matchy")

    conn =
      @good_payload
      |> setup_conn(@good_signature)
      |> call(opts)

    refute :sanity_webhook_error in Map.keys(conn.private)

    opts = Keyword.put(@opts, :path, ["/no-matchy", "/no-matchy-2"])

    conn =
      @good_payload
      |> setup_conn(@good_signature)
      |> call(opts)

    refute :sanity_webhook_error in Map.keys(conn.private)
  end

  test "computes signature" do
    assert {:ok, computed} =
             SanityWebhookPlug.Signature.compute(@good_ts, @good_payload, @opts[:secret])

    assert computed == @good_hash

    conn =
      @good_payload
      |> setup_conn(@good_signature)
      |> call(@opts)

    debug = conn.private.sanity_webhook_plug
    assert debug.error == false
    assert debug.computed == computed
    assert debug.hash == @good_hash
    assert debug.ts == @good_ts
    assert debug.body == nil
    assert debug.secret == @opts[:secret]
  end

  test "reads secret from MFA" do
    opts = Keyword.put(@opts, :secret, {__MODULE__, :bad_secret, []})

    conn =
      @good_payload
      |> setup_conn(@good_signature)
      |> call(opts)

    expected_message = "Sanity webhook signature does not match expected"
    assert conn.resp_body == Jason.encode!(%{error: expected_message})
    assert %{error: ^expected_message} = conn.private.sanity_webhook_plug

    # --

    opts = Keyword.put(@opts, :secret, &__MODULE__.error_secret/0)

    conn =
      @good_payload
      |> setup_conn(@good_signature)
      |> call(opts)

    expected_message = "no secret"
    assert conn.resp_body == Jason.encode!(%{error: expected_message})
    assert %{error: ^expected_message} = conn.private.sanity_webhook_plug

    # --

    opts = Keyword.put(@opts, :secret, {__MODULE__, :secret, []})

    conn =
      @good_payload
      |> setup_conn(@good_signature)
      |> call(opts)

    assert %{error: false} = conn.private.sanity_webhook_plug

    # --

    opts = Keyword.put(@opts, :secret, {__MODULE__, :bare_secret, []})

    conn =
      @good_payload
      |> setup_conn(@good_signature)
      |> call(opts)

    assert %{error: false} = conn.private.sanity_webhook_plug
  end

  test "halts on incorrect signature - timestamp" do
    bad_ts = 1_633_519_811_999
    bad_signature = "t=#{bad_ts},v1=#{@good_hash}"

    conn =
      @good_payload
      |> setup_conn(bad_signature)
      |> call(@opts)

    expected_message = "Sanity webhook signature does not match expected"
    assert conn.resp_body == Jason.encode!(%{error: expected_message})

    debug = conn.private.sanity_webhook_plug
    assert debug.error == expected_message
    assert debug.computed == "MvOplWzHD4SnEHitPZJmur5XzUATpQdN4oFX1ndiW7g"
    assert debug.hash == @good_hash
    assert debug.computed != debug.hash
    assert debug.ts == bad_ts
    assert debug.body == @good_payload
    assert debug.secret == @opts[:secret]
  end

  test "halts on incorrect signature - old timestamp" do
    bad_ts = 123
    bad_signature = "t=#{bad_ts},v1=#{@good_hash}"

    conn =
      @good_payload
      |> setup_conn(bad_signature)
      |> call(@opts)

    expected_message = "Timestamp 123 is too early to be a valid Sanity webhook"
    assert %{error: ^expected_message} = conn.private.sanity_webhook_plug
    assert conn.resp_body == Jason.encode!(%{error: expected_message})
  end

  test "halts on incorrect signature - hash" do
    bad_signature = "t=#{@good_ts},v1=badhash"

    conn =
      @good_payload
      |> setup_conn(bad_signature)
      |> call(@opts)

    expected_message = "Sanity webhook signature does not match expected"
    assert %{error: ^expected_message} = conn.private.sanity_webhook_plug
    assert conn.resp_body == Jason.encode!(%{error: expected_message})
  end

  test "halts on incorrect signature - payload" do
    bad_payload = "{\"_id\":\"foo\"}"

    conn =
      bad_payload
      |> setup_conn(@good_signature)
      |> call(@opts)

    expected_message = "Sanity webhook signature does not match expected"
    assert %{error: ^expected_message} = conn.private.sanity_webhook_plug
    assert conn.resp_body == Jason.encode!(%{error: expected_message})
  end

  test "json decoding errors" do
    binary = <<0, 1, 2, 3, 4, 5, 6, 7, 8, 9>>
    {:ok, hash} = Signature.compute(@good_ts, binary, @opts[:secret])
    signature = "t=#{@good_ts},v1=#{hash}"

    conn =
      binary
      |> setup_conn(signature)
      |> call(@opts)

    assert %{error: %Jason.DecodeError{}} = conn.private.sanity_webhook_plug
    assert conn.resp_body == Jason.encode!(%{error: "error"})
  end

  test "skips when already halted" do
    conn =
      @good_payload
      |> setup_conn(@good_signature)
      |> Plug.Conn.halt()
      |> call(@opts)

    assert conn.halted
    refute SanityWebhookPlug.get_debug(conn)
  end

  # test "works with smaller read lengths" do
  #   opts = Keyword.put(@opts, :length, 8)
  #
  #   conn =
  #     @good_payload
  #     |> setup_conn(@good_signature)
  #     |> call(opts)
  # end

  test "halts on incorrect header" do
    conn =
      :post
      |> conn("/sanity", @good_payload)
      |> put_req_header("some-other-header", @good_signature)
      |> put_req_header("content-type", "application/json")
      |> put_req_header("user-agent", "Sanity.io webhook delivery")
      |> put_req_header("accept-encoding", "gzip")
      |> call(@opts)

    expected_message = "Could not find valid Sanity webhook signature header"
    assert %{error: ^expected_message} = conn.private.sanity_webhook_plug
    assert conn.resp_body == Jason.encode!(%{error: expected_message})
  end

  describe "make_header" do
    test "generates a signature tuple" do
      dt = DateTime.new!(~D[2023-01-01], ~T[00:00:00], "Etc/UTC")
      body = "foo"
      secret = bare_secret()

      assert SanityWebhookPlug.make_header(dt, body, secret) ==
               {"sanity-webhook-signature",
                "t=1672531200000,v1=dMWBgdDyejVsuXjFpQD-nMtulFv1pE2BBz5cNNzINcA"}

      dt = 1_672_531_200_000

      assert SanityWebhookPlug.make_header(dt, body, secret) ==
               {"sanity-webhook-signature",
                "t=1672531200000,v1=dMWBgdDyejVsuXjFpQD-nMtulFv1pE2BBz5cNNzINcA"}

      invalid_dt = 123

      assert {:error, _, _} = SanityWebhookPlug.make_header(invalid_dt, body, secret)
    end
  end

  defp setup_conn(payload, signature) do
    :post
    |> conn("/sanity?foo=bar", payload)
    |> put_req_header(SanityWebhookPlug.header(), signature)
    |> put_req_header("content-type", "application/json")
    |> put_req_header("user-agent", "Sanity.io webhook delivery")
    |> put_req_header("accept-encoding", "gzip")
  end

  defp call(conn, opts) do
    SanityWebhookPlug.call(conn, SanityWebhookPlug.init(opts))
  end
end


mix.exs
defmodule SanityWebhookPlug.MixProject do
  use Mix.Project
  @version "0.1.4"

  def project do
    [
      app: :sanity_webhook_plug,
      version: @version,
      elixir: "~> 1.14",
      preferred_cli_env: [dialyzer: :test, credo: :test, tests: :test],
      start_permanent: Mix.env() == :prod,
      aliases: aliases(),
      package: package(),
      deps: deps(),
      docs: docs(),
      homepage_url: "https://hexdocs.pm/sanity_webhook_plug",
      source_url: "https://github.com/dbernheisel/sanity_webhook_plug",
      description: "Plug for handling Sanity.io GROQ-powered webhooks"
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger]
    ]
  end

  defp docs do
    [
      source_ref: @version,
      extras: ["CHANGELOG.md", "LICENSE"]
    ]
  end

  defp package do
    %{
      maintainers: ["David Bernheisel"],
      licenses: ["Apache-2.0"],
      files: [
        "lib",
        "mix.exs",
        "CHANGELOG*",
        "README*",
        "LICENSE*"
      ],
      links: %{
        "GitHub" => "https://github.com/dbernheisel/sanity_webhook_plug",
        "Readme" => "https://github.com/dbernheisel/sanity_webhook_plug/blob/#{@version}/README.md",
        "Changelog" =>
          "https://github.com/dbernheisel/sanity_webhook_plug/blob/#{@version}/CHANGELOG.md",
        "Sanity Webhook Docs" => "https://www.sanity.io/docs/webhooks"
      }
    }
  end

  defp deps do
    [
      {:plug, "~> 1.0"},
      # dev/test
      {:credo, "~> 1.7", only: [:dev, :test], runtime: false},
      {:jason, "~> 1.0", only: [:dev, :test]},
      {:dialyxir, "~> 1.0", only: [:dev, :test], runtime: false},
      {:ex_doc, ">= 0.0.0", only: :dev, runtime: false}
    ]
  end

  defp aliases do
    [
      tests: ["test", "dialyzer", "credo", "format --check-formatted"]
    ]
  end
end


lib/sanity_webhook_plug/signature.ex
defmodule SanityWebhookPlug.Signature do
  @moduledoc """
  Compute and verify signatures from Sanity webhooks
  """

  @minimum_ts 1_609_459_200_000

  @type secret :: mfa() | (() -> String.t() | {:ok, String.t()}) | String.t()

  @doc """
  Verify a payload, timestamp, and secret against a computed signature
  """
  @spec verify(String.t(), pos_integer(), binary(), secret()) ::
          :ok | {:error, String.t(), String.t()}
  def verify(hash, ts, payload, secret) do
    case compute(ts, payload, secret) do
      {:ok, computed} ->
        if Plug.Crypto.secure_compare(hash, computed) do
          :ok
        else
          {:error, "Sanity webhook signature does not match expected", computed}
        end

      error ->
        error
    end
  end

  @doc """
  Compute the signature for Sanity webhooks
  """
  @spec compute(pos_integer(), binary(), String.t()) ::
          {:ok, String.t()} | {:error, String.t(), nil}
  def compute(ts, _payload, _secret) when ts < @minimum_ts,
    do: {:error, "Timestamp #{ts} is too early to be a valid Sanity webhook", nil}

  def compute(ts, payload, secret) do
    {:ok, :hmac |> :crypto.mac(:sha256, secret, "#{ts}.#{payload}") |> base64url_encode()}
  end

  def base64url_encode(payload) do
    payload
    |> Base.encode64(padding: false)
    |> String.replace("+", "-")
    |> String.replace("/", "_")
  end

  def base64url_decode(payload) do
    payload
    |> String.replace("_", "/")
    |> String.replace("-", "+")
    |> Base.decode64!(padding: false)
  end
end


lib/sanity_webhook_plug.ex
defmodule SanityWebhookPlug do
  @external_resource "README.md"
  @moduledoc "README.md"
             |> File.read!()
             |> String.split("<!-- MDOC !-->")
             |> Enum.fetch!(1)

  @behaviour Plug
  require Logger
  alias SanityWebhookPlug.Signature

  @derive {Inspect, except: [:secret]}

  @type t :: %__MODULE__{
          body: binary() | nil,
          computed: String.t(),
          error: String.t() | false,
          secret: String.t(),
          hash: String.t(),
          ts: pos_integer()
        }

  defstruct [:body, :computed, :secret, :hash, :ts, error: false]

  @header "sanity-webhook-signature"
  @plug_key :sanity_webhook_plug

  @doc """
  Initialize SanityWebhookPlug options.
  """
  def init(opts) do
    path_info = opts |> Keyword.fetch!(:at) |> String.split("/", trim: true)
    handler = Keyword.fetch!(opts, :handler)

    sanity_json = Keyword.get(opts, :json_decoder)
    phoenix_json = Application.get_env(:phoenix, :json_library)
    jason = if Code.ensure_loaded?(Jason), do: Jason
    poison = if Code.ensure_loaded?(Poison), do: Poison

    [
      path_info,
      handler,
      Keyword.get(opts, :secret),
      Keyword.take(opts, [:length, :read_length, :read_timeout]),
      sanity_json || phoenix_json || jason || poison
    ]
  end

  @doc """
  Process the conn for a Sanity Webhook and verify its authenticity.
  """
  def call(conn, opts)

  def call(%Plug.Conn{halted: true} = conn, _), do: conn

  def call(%Plug.Conn{path_info: path_info} = conn, [path_info | opts]) do
    call(conn, opts)
  end

  def call(conn, [handler, secret, read_opts, json_mod]) do
    with {:ok, secret} <- get_secret(secret),
         {:ok, conn, body} <- read_body(conn, read_opts),
         {:ok, conn, {ts, hash}} <- get_signature(conn),
         :ok <- verify(conn, hash, ts, body, secret),
         {:ok, json} <- parse(json_mod, body, conn, ts, hash) do
      conn
      |> put_debug({hash, ts, nil}, hash, secret, false)
      |> handler.handle_event(json)
      |> Plug.Conn.halt()
    else
      {:error, error} ->
        # Bad secret or bad JSON decoding
        conn
        |> put_debug(nil, nil, secret, error)
        |> handle_error()
        |> handler.handle_error(error)

      {:error, error, conn} ->
        # Bad body read or no header
        conn
        |> put_debug(nil, nil, secret, error)
        |> handle_error()
        |> handler.handle_error(error)

      {:error, error, conn, components, computed} ->
        # Bad signature
        conn
        |> put_debug(components, computed, secret, error)
        |> handle_error()
        |> handler.handle_error(error)
    end
  end

  def call(conn, _opts), do: conn

  defp parse(json_mod, body, conn, ts, hash) do
    case json_mod.decode(body) do
      {:ok, json} ->
        {:ok, Map.merge(json, conn.query_params)}

      {:error, error} ->
        {:error, error, conn, {ts, hash, body}, nil}
    end
  end

  @doc """
  Get the Sanity Webhook debug information from the conn.
  """
  @spec get_debug(Plug.Conn.t()) :: t()
  def get_debug(conn), do: conn.private[@plug_key]

  @doc """
  Generate a request header tuple compatible with Sanity webhook systems.

  Timestamp can be either a %DateTime{} or a unix timestamp in milliseconds.

      {"sanity-webhook-signature", "ts=123,v1=abc123signature"}
  """
  @spec make_header(DateTime.t() | pos_integer(), binary(), String.t()) ::
          {String.t(), String.t()} | {:error, String.t(), nil}
  def make_header(%DateTime{} = timestamp, payload, secret) do
    make_header(DateTime.to_unix(timestamp, :millisecond), payload, secret)
  end

  def make_header(timestamp, payload, secret) when is_integer(timestamp) do
    with {:ok, sig} <- Signature.compute(timestamp, payload, secret) do
      {header(), "t=#{timestamp},v1=#{sig}"}
    end
  end

  @doc """
  The expected request header that contains the Sanity webhook signature and timestamp
  """
  @spec header() :: String.t()
  def header, do: @header

  defp verify(conn, hash, ts, body, secret) do
    case Signature.verify(hash, ts, body, secret) do
      {:error, message, computed} -> {:error, message, conn, {hash, ts, body}, computed}
      ok -> ok
    end
  end

  defp read_body(%{body_params: %Plug.Conn.Unfetched{}} = conn, read_opts) do
    read_body(conn, "", Plug.Conn.read_body(conn, read_opts), read_opts)
  end

  defp read_body(conn, _read_opts), do: {:ok, conn, conn.body_params}

  defp read_body(_conn, body, {:ok, more_body, conn}, read_opts) do
    {:ok, Plug.Conn.fetch_query_params(conn, read_opts), body <> more_body}
  end

  defp read_body(_conn, body, {:more, more_body, conn}, read_opts) do
    read_body(conn, body <> more_body, Plug.Conn.read_body(conn, read_opts), read_opts)
  end

  defp read_body(conn, _body, {:error, error}, _read_opts), do: {:error, error, conn}

  defp get_signature(conn) do
    case Plug.Conn.get_req_header(conn, @header) do
      [header] ->
        %{"ts" => ts, "v1" => hash} =
          Regex.named_captures(~r/^t=(?<ts>\d+)[, ]v1=(?<v1>[^, ]+)$/, String.trim(header))

        {:ok, conn, {String.to_integer(ts), String.trim(hash)}}

      _ ->
        {:error, "Could not find valid Sanity webhook signature header", conn}
    end
  end

  defp put_debug(conn, nil, nil, secret, error) do
    secret = secret |> get_secret() |> elem(1)
    Plug.Conn.put_private(conn, @plug_key, %__MODULE__{error: error, secret: secret})
  end

  defp put_debug(conn, {hash, ts, body}, computed, secret, error) do
    {:ok, secret} = get_secret(secret)

    Plug.Conn.put_private(conn, @plug_key, %__MODULE__{
      hash: hash,
      error: error,
      ts: ts,
      computed: computed,
      secret: secret,
      body: body
    })
  end

  defp handle_error(conn), do: conn |> Plug.Conn.put_status(400) |> Plug.Conn.halt()

  defp get_secret({m, f, a}), do: get_secret(apply(m, f, a))
  defp get_secret(fun) when is_function(fun), do: get_secret(fun.())
  defp get_secret({:ok, secret}) when is_binary(secret), do: {:ok, secret}
  defp get_secret({:ok, nil}), do: get_secret(nil)
  defp get_secret({:error, _} = error), do: error
  defp get_secret(secret) when is_binary(secret), do: {:ok, secret}

  defp get_secret(nil) do
    case Application.get_env(:sanity_webhook_plug, :webhook_secret) do
      nil ->
        {:error, "No secret configured for SanityWebhookPlug"}

      secret ->
        {:ok, secret}
    end
  end
end
